ui/vr_webgl.js

```diff
@@ -0,0 +1,611 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+goog.provide('shaka.ui.VRWebgl');
+
+goog.require('goog.asserts');
+goog.require('shaka.log');
+goog.require('shaka.Player');
+goog.require('shaka.ui.Matrix4x4');
+goog.require('shaka.ui.MatrixQuaternion');
+goog.require('shaka.ui.VRUtils');
+goog.require('shaka.util.EventManager');
+goog.require('shaka.util.IReleasable');
+goog.require('shaka.util.Timer');
+
+
+/**
+ * @implements {shaka.util.IReleasable}
+ */
+shaka.ui.VRWebgl = class {
+  /**
+   * @param {!HTMLMediaElement} video
+   * @param {!shaka.Player} player
+   * @param {!HTMLCanvasElement} canvas
+   * @param {WebGLRenderingContext} gl
+   */
+  constructor(video, player, canvas, gl) {
+    /** @private {HTMLVideoElement} */
+    this.video_ = /** @type {!HTMLVideoElement} */ (video);
+
+    /** @private {shaka.Player} */
+    this.player_ = player;
+
+    /** @private {HTMLCanvasElement} */
+    this.canvas_ = canvas;
+
+    /** @private {WebGLRenderingContext} */
+    this.gl_ = gl;
+
+    /** @private {shaka.util.EventManager} */
+    this.eventManager_ = new shaka.util.EventManager();
+
+    /** @private {!Float32Array} */
+    this.originalQuaternion_ = shaka.ui.MatrixQuaternion.create();
+
+    /** @private {!Float32Array} */
+    this.currentQuaternion_ = shaka.ui.MatrixQuaternion.create();
+
+    this.shaderProgram_ = null;
+    this.verticesBuffer_ = null;
+    this.normalsBuffer_ = null;
+    this.verticesTextureCoordBuffer_ = null;
+    this.verticesIndexBuffer_ = null;
+
+    /** @private {!Float32Array} */
+    this.viewMatrix_ = shaka.ui.Matrix4x4.create();
+
+    /** @private {!Float32Array} */
+    this.projectionMatrix_ = shaka.ui.Matrix4x4.create();
+
+    /** @private {!Float32Array} */
+    this.normalMatrix_ = shaka.ui.Matrix4x4.create();
+
+    /** @private {!Float32Array} */
+    this.viewProjectionMatrix_ = shaka.ui.Matrix4x4.create();
+
+    /** @private {!Float32Array} */
+    this.identityMatrix_ = shaka.ui.Matrix4x4.create();
+
+    /** @private {?Float32Array} */
+    this.diff_ = null;
+
+    /** @private {boolean} */
+    this.stereoscopicMode_ = false;
+
+    /** @private {?shaka.util.Timer} */
+    this.activeTimer_ = null;
+
+    /** @private {number} */
+    this.previousCanvasWidth_ = 0;
+
+    /** @private {number} */
+    this.previousCanvasHeight_ = 0;
+
+    /**
+     * @private {?{vertices: !Array.<number>, texCoords: !Array.<number>,
+     *           indices: !Array.<number>}}
+     */
+    this.sphere_ = null;
+
+    /** @private {?number} */
+    this.vertexPositionAttribute_ = null;
+
+    /** @private {?number} */
+    this.textureCoordAttribute_ = null;
+
+    /** @private {?WebGLTexture} */
+    this.texture_ = null;
+
+    /** @private {number} */
+    this.positionY_ = 0;
+
+    /** @private {number} */
+    this.fieldOfView_ = 75;
+
+    /** @private {number} */
+    this.animDuration_ = 0.5;
```

Comment:
It looks like this value never changes. Why make it a variable and not a constant?

Reply:
Done!